src/qat/ir/instructions.py

```diff
@@ -0,0 +1,1032 @@
+    another form of component.
+    """
+
+    inst: Literal["QuantumInstruction"] = "QuantumInstruction"
+    quantum_targets: Union[set[str], str]
+
+    @property
+    def duration(self):
+        return 0.0
+
+    @field_validator("quantum_targets", mode="before")
+    @classmethod
+    def _components_to_ids(cls, targets):
+        """
+        Fetches the IDs from quantum components.
+        """
+        if isinstance(targets, (set, list, tuple)):
+            targets = set(
+                [
+                    itm.full_id() if isinstance(itm, QuantumComponent) else itm
+                    for itm in targets
+                ]
+            )
+            targets = [val for val in targets][0] if len(targets) == 1 else targets
+        else:
+            targets = (
+                targets.full_id() if isinstance(targets, QuantumComponent) else targets
+            )
+        return targets
+
+
+class PhaseShift(QuantumInstruction):
+    """
+    A PhaseShift instruction is used to change the phase of waveforms sent down
+    the pulse channel.
+    """
+
+    inst: Literal["PhaseShift"] = "PhaseShift"
+    quantum_targets: str
+    phase: Union[float, Variable] = 0.0
+
+    def __init__(self, quantum_targets: PulseChannel, phase: float, **kwargs):
+        super().__init__(quantum_targets=quantum_targets, phase=phase)
+
+    @property
+    def channel(self):
+        return self.quantum_targets
+
+    def __repr__(self):
+        return f"phaseshift {self.channel},{self.phase}"
+
+    @field_validator("quantum_targets", mode="before")
+    @classmethod
+    def _is_pulse_channel(cls, target):
+        if not isinstance(target, (str, PulseChannel)):
+            raise ValueError(
+                f"channel has type {str(type(target))}: it must have type PulseChannel"
+            )
+        return target.full_id() if isinstance(target, PulseChannel) else target
+
+
+class FrequencyShift(QuantumInstruction):
+    """Change the frequency of a pulse channel."""
+
+    inst: Literal["FrequencyShift"] = "FrequencyShift"
+    quantum_targets: str
+    frequency: Union[float, Variable] = 0.0
+
+    def __init__(self, quantum_targets: PulseChannel, frequency: float, **kwargs):
+        super().__init__(quantum_targets=quantum_targets, frequency=frequency)
+
+    @property
+    def channel(self):
+        return self.quantum_targets
+
+    def __repr__(self):
+        return f"frequencyshift {self.channel},{self.frequency}"
+
+    @field_validator("quantum_targets", mode="before")
+    @classmethod
+    def _is_pulse_channel(cls, target):
+        if not isinstance(target, (str, PulseChannel)):
+            raise ValueError(
+                f"channel has type {str(type(target))}: it must have type PulseChannel"
+            )
+        return target.full_id() if isinstance(target, PulseChannel) else target
+
+
+class Id(QuantumInstruction):
+    """Simply a no-op, called an Identity gate."""
+
+    # TODO: do we really need this?
```

Comment:
Sorry I'm not sure what you mean :sweat_smile: 

My comment just meant do we really need an instruction that does nothing? `Delay` tells a particular channel to do nothing for some time, this just seems like some mixing of "gates" and "instructions". I think it can probably go

Reply:
Oh I misunderstood the purpose of this class, thought it was a id tag. Anyway, it can be nice to have as a conceptual gate to improve the clarity for some circuits, I don't think it does any harm anyway